### ISSUE TITLE
Implement `axis_size` and `axis_index` like jax.lax.axis_index

### ISSUE THREAD

**Issue by fmassa** [meta-pytorch/autoparallel#165]
When using `local_map`, it might be important to know the size of the mesh on a given axis, as well as the location where the current program is.
This is needed e.g. when implementing context parallel, as the masking of the query depends on where in the sequence we are, which depends on the size and location of the GPU in the mesh.

One workaround would be to use `torch.distributed.get_rank()` and `torch.distributed.get_world_size()`, but this has two issues:
- we get instead a scalar baked into the graph, which might worth but will break specially if we compile on a single rank
- we need to pass a process group if we want to get a specific mesh dim size, but we can't have process groups in the graph.

API references for JAX:
- `axis_index` https://docs.jax.dev/en/latest/_autosummary/jax.lax.axis_index.html
- `axis_size` https://docs.jax.dev/en/latest/_autosummary/jax.lax.axis_size.html

Ultimately, this functionality should be in PyTorch though, but I'm creating it here for easier tracking as we need it quickly for AutoParallel
